core/sail/lucene/src/main/java/org/eclipse/rdf4j/sail/lucene/LuceneDocument.java

```diff
@@ -140,13 +154,33 @@ public boolean hasProperty(String fieldName, String value) {
 	public void addGeoProperty(String field, String value) {
 		LuceneIndex.addStoredOnlyPredicateField(field, value, doc);
 		try {
-			SpatialStrategy geoStrategy = geoStrategyMapper.apply(field);
-			Shape shape = geoStrategy.getSpatialContext().readShapeFromWkt(value);
-			for (IndexableField f : geoStrategy.createIndexableFields(shape)) {
-				doc.add(f);
+			String wkt = value;
+			// wkt = wkt.replace("\"", "").replace("^^<http://www.opengis.net/ont/geosparql#wktLiteral>", "");
+			// System.out.println(wkt);
+			Object shape = SimpleWKTShapeParser.parse(wkt);
+			if (shape instanceof Polygon[]) {
+				for (Polygon p : (Polygon[]) shape) {
+					for (Field f : LatLonShape.createIndexableFields(field, p)) {
+						doc.add(f);
+					}
+				}
+			} else if (shape instanceof Polygon) {
+				for (Field f : LatLonShape.createIndexableFields(field, (Polygon) shape)) {
+					doc.add(f);
+				}
+			} else if (shape instanceof Line) {
+				for (Field f : LatLonShape.createIndexableFields(field, (Line) shape)) {
+					doc.add(f);
+				}
+			} else if (shape instanceof double[]) {
+				double point[] = (double[]) shape;
+				// System.out.println(point[0]+" "+point[1]);
+				doc.add(new LatLonPoint(field, point[1], point[0]));
+			} else {
+				throw new IllegalArgumentException("Geometry is not supported");
```

Comment:
Could we add the shape object to the exception message? Not sure it has a readable string representation but on the off chance, it could help pinpoint the error.

Reply:
More generally: I don't know the geospatial support well enough to really follow, but I'm a little uncomfortable to have this if..else construction in our code base. Here, we make a choice to just support Polygon, Line, and coordinate tuples as possible Shapes, and everything else we just throw an exception. Are there no other possible shapes that could come from processing the WKT literal that could be indexed by Lucene? For example what about multiline shapes, or rectangles/envelopes, or more generally Geometry collections?

Reply:
Yes you are totally right, Lucene could parse all those types:("MULTIPOLYGON","POLYGON","POINT","MULTIPOINT","LINESTRING","MULTILINESTRING","ENVELOPE","GEOMETRYCOLLECTION"). 
I can support those extra shapes at index time but what would you suggest to remove this if else construction ? Is there a specific style you prefer so that I can follow ?

Reply:
I had a closer look at how createIndexableFields works and I guess we're stuck with the approach. It's somewhat inevitable since the output of SimpleWKTShapeParser is an Object - so the core of the problem is not in our code but in how that parser's api is designed. 

That's a long-winded way of saying that I'm fine with sticking with the if-else construction :)

Reply:
At some point this logic has to be done somewhere, in the current implementation in spatial4j, it was done by inherting the Shape class to any other geometry and at index time the indexer wouldn't care about the type. But with the Lucene parser and the new indexing scenario there should some logic because (shapes are not indexed with points.. etc). I have changed the code again based on your demand to cover all the available WKT geomerties, and even at search time I did some modifications for the new parser,indexer and searcher.. I will commit the modifications soon.